casper/src/test/scala/coop/rchain/casper/batch2/BlockReceiverEffectsSpec.scala

```diff
@@ -0,0 +1,243 @@
+          bs.contains(Seq(block.blockHash)) wasCalled once
+          br.ackReceived(block.blockHash) wasCalled once
+          dagStorageWasNotModified(bds)
+          outList shouldBe List(block.blockHash)
+        }
+    }
+
+  // Provided to BlockReceiver shard name ("test") is differ from block's shard name ("root" by default)
+  // So block should be rejected and output stream should never take block
+  it should "discard block with invalid shard name" in withEnv("test") {
+    case (incomingQueue, _, outStream, bs, br, bds) =>
+      for {
+        block <- makeBlock()
+        _     <- incomingQueue.enqueue1(block)
+      } yield {
+        bs.put(*) wasNever called
+        bs.contains(*) wasNever called
+        br.ackReceived(*) wasNever called
+        dagStorageWasNotModified(bds)
+        outStream should notEmit
+      }
+  }
+
+  it should "discard block with invalid block hash" in withEnv("root") {
+    case (incomingQueue, _, outStream, bs, br, bds) =>
+      for {
+        block <- makeBlock().map(_.copy(blockHash = "abc".unsafeHexToByteString))
+        _     <- incomingQueue.enqueue1(block)
+      } yield {
+        bs.put(*) wasNever called
+        bs.contains(*) wasNever called
+        br.ackReceived(*) wasNever called
+        dagStorageWasNotModified(bds)
+        outStream should notEmit
+      }
+  }
+
+  it should "discard block with invalid signature" in withEnv("root") {
+    case (incomingQueue, _, outStream, bs, br, bds) =>
+      for {
+        block <- makeBlock().map(_.copy(sig = "abc".unsafeHexToByteString))
+        _     <- incomingQueue.enqueue1(block)
+      } yield {
+        bs.put(*) wasNever called
+        bs.contains(*) wasNever called
+        br.ackReceived(*) wasNever called
+        dagStorageWasNotModified(bds)
+        outStream should notEmit
+      }
+  }
+
+  it should "discard known block" in withEnv("root") {
+    case (incomingQueue, _, outStream, bs, br, bds) =>
+      for {
+        block <- addBlock(bs)
+        _     <- incomingQueue.enqueue1(block)
+      } yield {
+        bs.put(Seq((block.blockHash, block))) wasCalled once
+        bs.contains(*) wasNever called
+        br.ackReceived(*) wasNever called
+        dagStorageWasNotModified(bds)
+        outStream should notEmit
+      }
+  }
+
+  it should "pass to output blocks with resolved dependencies" in withEnv("root") {
+    case (incomingQueue, validatedQueue, outStream, bs, br, bds) =>
+      for {
+        // Received a parent with an empty list of justifications and its child
+        a1 <- makeBlock()
+        a2 <- makeBlock(List(a1.blockHash))
+
+        // Put the parent and child in the input queue
+        _ <- incomingQueue.enqueue1(a2)
+        _ <- incomingQueue.enqueue1(a1)
+
+        // Dependencies of the child (its parent) have not yet been resolved,
+        // so only the parent goes to the output queue, since it has no dependencies
+        a1InOutQueue <- outStream.take(1).compile.toList.map(_.head)
```

Comment:
There is `lastOrError` function which returns only one element, in this case head and last should be the same.